### PR TITLE
Boost drawing speed of ItemLists with many items

### PR DESCRIPTION
**Pull request by ghost**
With 2000+ items, `ItemList` (used for displaying errors, messages) starts to get really sluggish, and it starts to hog the whole Godot UI for larger numbers.

The current implementation draws every row (including the non-visible ones) on every frame due to a strange clip rect.

This PR polishes the whole drawing code, such that lists of several 1000 elements scroll smoothly now.

### CODE REVIEW

**Comment by Geequlim** [comment]
@poke1024 How about the Tree? Does it need such an optimization？

**Comment by poke1024** [comment]
@Geequlim `Tree` should be ok; even though it traverses the whole tree on draw, it doesn't send any content that is invisible to the visual server (which is the performance problem in `ItemList`). The only code I see in `Tree` that does not scale at all is the drawing of the relationship lines; I'll do a small PR.

File: scene/gui/item_list.cpp [hunk 1]
```diff
@@ -960,12 +960,30 @@ void ItemList::_notification(int p_what) {
 		Vector2 base_ofs = bg->get_offset();
 		base_ofs.y -= int(scroll_bar->get_value());
 
-		Rect2 clip(Point2(), size - bg->get_minimum_size() + Vector2(0, scroll_bar->get_value()));
+		Rect2 clip(-base_ofs, size);
+
+		int lo = 0;
```
Comment (akien-mga) [inline scene/gui/item_list.cpp]: Maybe a quick comment explaining what the lo/hi search does would be useful? The previous code was relatively self explanatory as it just clips and draws stuff, but now it's a bit more involved :)

**Comment by akien-mga** [comment]
We are now entering a strict release freeze for Godot 3.0 and will only consider major bug fixes. We won't merge new features and enhancements anymore until 3.0 is released.

Moving this PR to the 3.1 milestone, to be reviewed once the release freeze is lifted. It could eventually be cherry-picked for a future 3.0.1 maintenance release if it doesn't change the user-facing APIs and doesn't compromise the engine's stability.

**Comment by Geequlim** [comment]
IMO the PopupMenu menu also need such an optimization. You can find there are too much draw calls when click the language option button in the godot project manager. A large part of that to draw the items which are not visiable.

Draw calls when the language option showing : 1251
Draw calls without the language option : 528

By the way one draw call for each character also needs an optimization.

**Comment by Geequlim** [comment]
IMO the PopupMenu menu also need such an optimization. You can find there are too much draw calls when click the language option button in the godot project manager. A large part of that to draw the items which are not visiable.

Draw calls when the language option showing : 1251
Draw calls without the language option : 528

By the way one draw call for each character also needs an optimization.

**Comment by poke1024** [comment]
> By the way one draw call for each character also needs an optimization.

I agree, that would be nice, together with a variant of `get_char_size` for multiple chars.

File: scene/gui/item_list.cpp (hunk 1 above)
Comment (poke1024) [reply scene/gui/item_list.cpp]: Added comments and cleaned up the var scopes; oh C++11 lambdas, where are you?

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**